Attach a new finalizer in GCE ILB creation.

prameshj opened the pull request:
This change adds a new finalizer to identify LoadBalancer services that are using GCE ILB and managed by service controller. Logic is also added in service_controller to skip create/update of services if a finalizer from a different controller is found.

The newly added finalizer will be checked by other controllers implementing ILB services to determine if a given service is already being managed by service_controller.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
This PR adds a new finalizer to identify LoadBalancer services that are using GCE ILB and managed by service controller.
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
/kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

prameshj commented:
/assign @MrHohn @bowei 
/sig network
/priority-important-soon

bowei commented on cluster/gce/addons/loadbalancing/cloud-provider-role.yaml:
```diff
@@ -33,6 +33,13 @@ rules:
   - patch
   - update
+- apiGroups:
+  - ""
```
what does empty mean here?

prameshj replied:
For the coreV1 apis, the apiGroup is left as empty in the existing PolicyRules. I followed the same convention here.

bowei commented on pkg/controller/service/controller.go:
```diff
@@ -334,6 +334,13 @@ func (s *Controller) syncLoadBalancerIfNeeded(service *v1.Service, key string) (
 	} else {
 		// Create or update the load balancer if service wants one.
 		op = ensureLoadBalancer
+		if !servicehelper.HasLBFinalizer(service) && len(service.ObjectMeta.Finalizers) > 0 {
```
This is pretty broad. We may want to think about what this means for users that use finalizers that we don't know about.

prameshj replied:
I agree. I didn't want to put checks for cloud-provider specific finalizer keys here either.
Any users that want service controller to manage the services would keep the default service controller finalizer. This check will be true if the default finalizer is NOT present and some other finalizer key is.

MrHohn replied:
Seems strange that adding a random finalizer could disable the entire code path. Does this have to live in the service controller layer? We can check explicitly if it is within gce provider.

MrHohn commented on pkg/controller/service/controller.go:
```diff
@@ -348,6 +355,7 @@ func (s *Controller) syncLoadBalancerIfNeeded(service *v1.Service, key string) (
 				// functionality is implemented by a different controller.  In this case, we
 				// return immediately without doing anything.
 				klog.V(4).Infof("LoadBalancer for service %s implemented by a different controller %s, Ignoring error", key, s.cloud.ProviderName())
+				s.removeFinalizer(service)
```
Check for error?

MrHohn replied:
Would leaving the default finalizer attached cause problem?

prameshj replied:
The finalizer removal was added so that the check in line 337 would work for services that are managed by a different controller. The condition is - load-balancer-cleanup finalizer should not be present AND an unrecognized finalizer should be present.

Your recommendation to check explicitly for the v2 finalizer in gce provider code is good, we can move the check there and leave the load-balancer-cleanup finalizer intact in all cases.

MrHohn commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -40,14 +40,26 @@ import (
-
+	if g.client != nil {
+		// client will be empty in certain unit tests.
+		err := addFinalizer(svc, g.client.CoreV1(), LegacyILBFinalizer)
```
nit: `if err := ...`

bowei commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -41,13 +41,32 @@ import (
 	}
-
+	if hasFinalizer(svc, ILBFinalizerV2) {
+		// another conreoller is handling the resources for this service.
```
spelling (controller)

bowei commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -41,13 +41,32 @@ import (
+			"Skipped Ensure load balancer as service contains '%s' finalizer.", ILBFinalizerV2)
+		return nil, cloudprovider.ImplementedElsewhere
+	}
+	if g.client != nil {
```
not sure we want to introduce branches like this just to support unit tests

bowei commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -297,6 +316,12 @@ func (g *Cloud) ensureInternalLoadBalancerDeleted(clusterName, clusterID string,
 	if err := g.ensureInternalInstanceGroupsDeleted(igName); err != nil && !isInUsedByError(err) {
 		return err
 	}
+	if g.client != nil {
```
same down here, mock client in the unit tests to avoid introducing branches for testing sake

prameshj replied:
ok. I had started on the approach to use the mock client. We will also need to change all tests to also create the service so that the client.Get() will return a non-nil value. I had held off on it since it will be a much longer diff. I can include it with this PR.

bowei commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_util.go:
```diff
@@ -316,3 +320,84 @@ func typeOfNetwork(network *compute.Network) netType {
 	return fmt.Sprintf("projects/%s/locations/%s", project, zoneOrRegion)
 }
+
+func addFinalizer(service *v1.Service, kubeClient v1core.CoreV1Interface, key string) error {
```
you need unit test for these functions.

prameshj replied:
done

MrHohn commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -41,12 +41,30 @@ import (
+		return nil, cloudprovider.ImplementedElsewhere
+	}
+	if hasFinalizer(svc, ILBFinalizerV2) {
+		// another controller is handling the resources for this service.
```
nit: Another...

prameshj replied:
done

MrHohn commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -41,12 +41,30 @@ import (
 		return nil, cloudprovider.ImplementedElsewhere
 	}
+	err := addFinalizer(svc, g.client.CoreV1(), LegacyILBFinalizer)
+	if err != nil {
```
nit: `if err := addFinalizer()...`

prameshj replied:
done

MrHohn commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_util.go:
```diff
@@ -316,3 +322,84 @@ func typeOfNetwork(network *compute.Network) netType {
+	return c.Services(oldSvc.Namespace).Patch(oldSvc.Name, types.StrategicMergePatchType, patchBytes, "status")
+}
+
+func getPatchBytes(oldSvc *v1.Service, newSvc *v1.Service) ([]byte, error) {
```
Any chance to reuse https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/service/patch.go? Probably pkg/controller/service/util? So that we don't need to duplicate the codes and tests.

prameshj replied:
I didn't do that in order to keep dependency on k/k to minimum in cloud-provider code. I have opened https://github.com/kubernetes/kubernetes/pull/87217 to move this to k8s.io/cloud-provider/service/helpers. Once that merges, we can use those methods here and in pkg/controller/service/controller.go

prameshj replied:
Added a TODO to use method from service/helpers once a new release includes #87217

MrHohn commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -41,12 +41,29 @@ import (
 	// Used to list instances in all states(RUNNING and other) - https://cloud.google.com/compute/docs/reference/rest/v1/instanceGroups/listInstances
 	allInstances = "ALL"
+	// LegacyILBFinalizer key is used to identify ILB services whose resources are managed by service controller.
+	LegacyILBFinalizer = "gke.networking.io/l4-ilb-v1"
```
nit: ILBFinalizerLegacy?

prameshj replied:
Changed to ILBFinalizerV1 to match the string value.

MrHohn commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -41,12 +41,29 @@ import (
+	}
+	if hasFinalizer(svc, ILBFinalizerV2) {
+		// Another controller is handling the resources for this service.
+		g.eventRecorder.Eventf(svc, v1.EventTypeNormal, "SkippingEnsureLoadBalancer",
```
Should the above condition block also emit event like this?

prameshj replied:
Good idea, added.

MrHohn commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal_test.go:
```diff
@@ -1464,3 +1554,49 @@ func TestEnsureInternalFirewallPortRanges(t *testing.T) {
 	}
 }
+
+func TestEnsureInternalLoadBalancerFinalizer(t *testing.T) {
```
Any test that verifies service with ILBFinalizerV2 is skipped?

prameshj replied:
Added a new test.

MrHohn reviewed (approved):
/lgtm
/approve

MrHohn commented:
/lgtm

prameshj commented:
/assign @bowei for approval of cluster/gce changes.

prameshj commented:
/test pull-kubernetes-e2e-gce

MrHohn commented:
/lgtm

MrHohn commented:
/lgtm

MrHohn commented:
/lgtm

bowei commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -41,12 +41,31 @@ import (
 		return nil, cloudprovider.ImplementedElsewhere
 	}
+	if err := addFinalizer(svc, g.client.CoreV1(), ILBFinalizerV1); err != nil {
+		klog.Errorf("Failed to attach finalizer '%s' on service %s/%s - %v", ILBFinalizerV1, svc.Namespace,
```
we can merge but don't break the line @ 80 for go source code

prameshj replied:
done

bowei commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal_test.go:
```diff
@@ -155,8 +156,11 @@ func TestEnsureInternalLoadBalancer(t *testing.T) {
 	require.NoError(t, err)
-
 	svc := fakeLoadbalancerService(string(LBTypeInternal))
+	svc, err = gce.client.CoreV1().Services(svc.Namespace).Create(svc)
```
Why do we create this service twice
What is createInternalLoadBalancer() doing that we need to call create again here

prameshj replied:
The service is created here so that it can be looked up with a services "Get" command. This is needed so that the code in createInternalLoadBalancer can lookup and patch the service. That code does not create the k8s service again.

bowei replied:
Looking at the test code, it's pretty confusing why this is done?

prameshj replied:
I modified the tests so that creation of k8s service happens within a util function. PTAL.

prameshj replied:
discussed offline, we will leave the test code to create the services inline.

prameshj reviewed (commented):
/assign @bowei

bowei commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal_test.go:
```diff
@@ -213,7 +219,10 @@ func TestEnsureInternalLoadBalancerWithExistingResources(t *testing.T) {
 	require.NoError(t, err)
-	svc := fakeLoadbalancerService(string(LBTypeInternal))
+	svc, err := createFakeLoadbalancerService(string(LBTypeInternal), nil, gce)
+	if err != nil {
```
this clashes with the use of require
if you want to use require, then this should be require.NoError(t, err)

prameshj replied:
This test file has a mix of require, assert and if checks for error. I am trying to use if checks in the newer changes I make to this file. Is that ok?

bowei replied:
can we make it consistent w/in a given test function; otherwise we will never get convergence
also, if we are moving away from `require`, should drop a comment to say where we want to go for future edits

bowei commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_util_test.go:
```diff
@@ -118,3 +119,38 @@ func TestFirewallToGcloudArgs(t *testing.T) {
 }
+
+// TestFinalizerUtils tests the add/remove and hasFinalizer methods.
+func TestFinalizerUtils(t *testing.T) {
```
Usually don't name tests after files or packages but the functions you are testing:

TestFinalizerUtils

prameshj replied:
Done

prameshj committed:
Attach a new finalizer in GCE ILB creation.

Add logic in service_controller to skip create/update
if finalizer from a different controller is found.

The newly added finalizer will be checked by other controllers
implementing ILB services to determine if a given service is
already being managed by service_controller.

Moved finalizer check into cloudprovider code.

added unit test to verify new finalizer.

Modified existing unit test to create a fake service so that
attach/remove finalizer step can be tested.

prameshj commented:
/test pull-kubernetes-e2e-kind-ipv6

prameshj commented:
/test pull-kubernetes-e2e-gce

bowei commented:
/lgtm
/approve

prameshj commented:
/test pull-kubernetes-e2e-gce
